docs/user/reference/pings/index.md

```diff
@@ -83,3 +86,178 @@ GleanPings.search.submit("performed");
+    assertEquals(Pings.searchReasonCodes.performed, reason);
+    assertEquals(5, Search.INSTANCE.defaultEngine.testGetValue());
+    validatorRun = true;
+});
+
+// Submit the ping.
+Pings.INSTANCE.search.submit(
+    Pings.INSTANCE.searchReasonCodes.performed
+);
+
+// Verify that the validator run.
+assertTrue(callbackWasCalled);
+```
+
+</div>
+
+<div data-lang="Swift" class="tab">
+
+```swift
+@testable import Glean
+
+// Record some data.
+Search.defaultEngine.add(5)
+
+// Instruct the ping API to validate the ping data.
+var validatorRun = false
+GleanMetrics.pings.shared.search.testBeforeNextSubmit { reason in
+    XCTAssertEqual(.performed, reason, "Unexpected reason for search ping submitted")
+    XCTAssertEqual(5, try Search.defaultEngine.testGetValue(), "Unexpected value for default engine in search ping")
+    validatorRun = true
+}
+
+// Submit the ping.
+GleanMetrics.Pings.shared.search.submit(
+    reason: .performed
+)
+
+// Verify that the validator run.
+XCTAssert(validatorRun, "Expected validator to be called by now.")
+```
+
+</div>
+
+<div data-lang="Python" class="tab">
+
+```Python
+from glean import load_metrics, load_pings
+
+pings = load_pings("pings.yaml")
+metrics = load_metrics("metrics.yaml")
+
+# Record some data.
+metrics.search.default_engine.add(5)
+
+# Need a mutable object and plain booleans are not.
+callback_was_called = [False]
+
+def check_custom_ping(reason):
+    assert reason == pings.search_reason_codes.PERFORMED
+    assert 5 == metrics.search.default_engine.test_get_value()
+    callback_was_called[0] = True
+
+# Instruct the ping API to validate the ping data.
+pings.search.test_before_next_submit(check_custom_ping)
+
+# Submit the ping.
+pings.search.submit(pings.search_reason_codes.PERFORMED)
+
+# Verify that the validator run.
+assert callback_was_called[0]
+```
+
+</div>
+
+<div data-lang="Rust" class="tab">
+
+```Rust
+use glean::Pings;
+use glean_metrics;
+
+// Record some data.
+search::default_engine.add(5);
+
+// Instruct the ping API to validate the ping data.
+pings::search.test_before_next_submit(move |reason| {
+    assert_eq!(pings::SearchReasonCodes::Performed, reason);
+    assert_eq!(5, search::default_engine.test_get_value(None).unwrap());
+});
+
+// Submit the ping.
+pings::search.submit(pings::SearchReasonCodes::Performed);
+```
+
+</div>
+
+<div data-lang="JavaScript" class="tab">
+
+```js
+import * as search from "./path/to/generated/files/search.js";
+import * as pings from "./path/to/generated/files/pings.js";
+
+// Record some data.
+search.defaultEngine.add(5);
+
+// Instruct the ping API to validate the ping data.
+let validatorRun = false;
+const p = pings.search.testBeforeNextSubmit(async reason => {
+  assert.strictEqual(reason, "performed");
+  assert.strictEqual(await search.defaultEngine.testGetValue(), 5);
+  validatorRun = true;
+});
+
+// Submit the ping.
+pings.search.submit("performed");
+// Wait for the validation to finish.
```

Comment:
I could be asking for this due to being traumatized by this API, but I think it might be worth clearly stating the limitations of this API on JavaScript. That can be a follow up no problem. This could even be a whole page on the JavaScript specific section honestly........................................ Yeah, definitely I am asking out of _trauma_.

Reply:
https://bugzilla.mozilla.org/show_bug.cgi?id=1716540